http-svr/xmlrpc_client.ml

```diff
@@ -96,6 +98,64 @@ let get_new_stunnel_id =
   let m = Mutex.create () in
   fun () -> Mutex.execute m (fun () -> incr counter; !counter)
 
+let run_watchdog timeout (fire_fn : unit -> unit) f (pout,pin) =
+  let fired = ref None in
+  let th = Thread.create (fun () ->
+    match Unix.select [pout] [] [] timeout with
+      | [x],[],[] ->
+        fired := Some false
+      | _,_,_ ->
+        fire_fn ();
+        fired := Some true) () in
+  let cancel_watchdog () = let _x : int = Unix.write pin "x" 0 1 in () in
+  let get_fired () = Thread.join th; !fired in
+  f cancel_watchdog get_fired
+
+let with_pipe f =
+  let pout,pin = Unix.pipe () in
+  finally
+    (fun () -> f (pout,pin))
+    (fun () -> Unix.close pout; Unix.close pin)
+
+let watchdog timeout pid f =
+ with_pipe
+   (run_watchdog
+      timeout
+      (fun () ->
+         StunnelDebug.warn "Watchdog fired: killing pid: %d" pid;
+         Unix.kill pid Sys.sigterm)
```

Comment:
How do we know this is the right PID and stunnel hasn't died already and the PID has been reused by something else? Does stunnel have a lockfile it keeps open with the pid that we can use to kill it without race conditions?

Reply:
Only really by construction - we do this pattern elsewhere in the code too. I really want to throw away all this code and use direct openssl bindings instead, but since this is for backporting that's not really an option here.